disnake/interactions/base.py

```diff
@@ -689,23 +733,35 @@ async def defer(self, *, ephemeral: bool = False, with_message: bool = False) ->
         if self._responded:
             raise InteractionResponded(self._parent)
 
-        defer_type: int = 0
-        data: Optional[Dict[str, Any]] = None
+        defer_type: Optional[InteractionResponseType] = None
+        data: Dict[str, Any] = {}
         parent = self._parent
-        if (
-            parent.type in (InteractionType.application_command, InteractionType.modal_submit)
-            or with_message
-        ):
-            defer_type = InteractionResponseType.deferred_channel_message.value
-            if ephemeral:
-                data = {"flags": MessageFlags.ephemeral.flag}
+        if with_message is MISSING:
+            if ephemeral is not MISSING:
+                with_message = True
+
+        if parent.type in (InteractionType.application_command, InteractionType.modal_submit):
+            defer_type = InteractionResponseType.deferred_channel_message
         elif parent.type is InteractionType.component:
-            defer_type = InteractionResponseType.deferred_message_update.value
+            if with_message:
+                defer_type = InteractionResponseType.deferred_channel_message
+            else:
+                defer_type = InteractionResponseType.deferred_message_update
+
```

Comment:
could raise here instead of silently dropping the call, if none of the types match

Reply:
Fair, I initially didn't want to make it a breaking change but it *really* should raise instead.